### PR TITLE
Added cxx language standard detection

### PR DESCRIPTION
**Pull request by MathiasMagnus**
Added the ability to detect the required cxx standard version on a per-target basis. This is done to allow building SYCL targets with a specific language requirement as simply as:
```CMake
cmake_minimum_required(VERSION 3.2.2)

project(SYCL_CXX14 LANGUAGES CXX)

# Include modules
set(CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/cmake/Modules)
include(FindComputeCpp)

add_executable(generic src/hello_generic.cpp)

set_target_properties(generic PROPERTIES CXX_STANDARD 14
                                         CXX_STANDARD_REQUIRED ON)

add_sycl_to_target(generic ${CMAKE_CURRENT_BINARY_DIR}
                   ${CMAKE_CURRENT_SOURCE_DIR}/src/hello_generic.cpp)
```

_On a separate note, I do plan on overhauling the CMake detection scripts of both ComputeCpp and triSYCL to make them work better together. I'll open an issue to demo what I had in mind (before I burn my time for something nobody is interested in)._

### CODE REVIEW

**Comment by MathiasMagnus** [comment]
I forgot to mention. Computecpp-sdk samples compile fine without specifying any language standard, as expected. Extra behavior kicks in only when user specifies a standard to be used.

**Comment by DuncanMcBain** [comment]
Hi @MathiasMagnus, thanks for your patch! I'd be interested to know if this has come up in your development process, because realistically for ComputeCpp only two language versions make sense - 11 and 14. The options provided by computecpp_info will set the language version to 14, unless otherwise specified on the command line, but I can't think of any scenario where setting C++14 will break code that would otherwise work. Similarly we cannot in any way support C++98, and while clang-3.9 (on which our compiler is based) has decent support for C++ 17 there's no guarantee of it working in or with SYCL code.

I would be interested to know if you've managed to use any C++17 with our compiler. As I say it is possible that things might "just work", but they could just as easily fail, unfortunately.

That being said, do you also envision this being useful for stopping use of features that are from a later standard? e.g. accidentally using auto lambdas or similar when you the code is supposed to be C++11 only?

There is a separate issue (#32) for improving the CMake, and I have promised myself I'll make that change one day. Additionally there's issue #14 which I would like to revisit, as it probably does make sense for the CMake to be available in the package (as much as it's been useful for us to be able to update FindComputeCpp.cmake as and when necessary). Please feel free to get involved with those - it's an area that I've neglected, as you can see by the age of the issues, but I will make it better at some point.

Thanks,
Duncan.

**Comment by MathiasMagnus** [comment]
We too generally opt-in for the highest possible language standard that is portable enough to not cause too many headaches. Currently that is C++14 (though MSVC is still lacking, last time I checked Boost::Hana still did not compile).

You are correct, it might be smarter to throw a hard error on C++98, even though no entry point of SYCL explicitly requires C++11. cl::sycl::queue::submit can make do with an ordinary functor, it isn't mandated to be a lambda. Still, the implementations of SYCL are certainly sprinkled with C++11 all over the place.

As for C++17, the only reason we do not rely on it just yet is because it is far from widely available, it is roughly half done in Clang 3.9, but you are right there are nice features already implemented. constexpr if, fold expressions (I am just about to bake something instead of it for C++14), inline variables, just to name the most useful ones. (constexpr lambdas and template argument deduction for class templates missing in 3.9, roughly the 2 missing C++17 features we'd wish to rely on.) We will surely play with these as well, though production code will still target C++14 for a few months.

I am aware of those two issues, and I very much wanted to shove every extra logic into an imported target, however that does not allow running scripts on the target to extract the info we do today with add_sycl_to_target. I was wondering whether ComputeCpp and triSYCL are mature enough to have actual CMake support, such that CUDA received in CMake 3.9. I don't see that either project had their compilation model baked to a point that "rigid" external tooling would be a good idea. Other than that, the scripts could still be improved.

File: cmake/Modules/FindComputeCpp.cmake [hunk 1]
```diff
@@ -193,6 +193,21 @@ function(__build_spir targetName sourceFile binaryDir fileCounter)
...
+  set(device_compiler_cxx_standard)
+  get_target_property(targetCxxStandard ${targetName} CXX_STANDARD)
+  if (targetCxxStandard MATCHES 17)
+    set(device_compiler_cxx_standard "-std=c++1z")
+  elseif (targetCxxStandard MATCHES 14)
+    set(device_compiler_cxx_standard "-std=c++14")
+  elseif (targetCxxStandard MATCHES 11)
+    set(device_compiler_cxx_standard "-std=c++11")
+  elseif (targetCxxStandard MATCHES 98)
+    message(FATAL_ERROR "SYCL implementations cannot be compiled using C++98")
```
Comment (keryell) [inline cmake/Modules/FindComputeCpp.cmake]: What about C++03 and some other weird versions?
I guess we do not care... :-)

Comment (MathiasMagnus) [reply cmake/Modules/FindComputeCpp.cmake]: These are the ones CMake knows.

https://cmake.org/cmake/help/v3.9/prop_tgt/CXX_STANDARD.html

I'm not discriminating anything here... ;)

**Comment by Ruyk** [comment]
Does this work with older CMake versions? We are currently supporting 3.2.2 and above. If we have to bump the required version of CMake for the SDK we will have to adapt the documentation accordingly...

**Comment by MathiasMagnus** [comment]
It does. I do not have an installation available, but the [online docs for 3.2](https://cmake.org/cmake/help/v3.2/prop_tgt/CXX_STANDARD.html?highlight=cxx_standard) already mention this variable. 17 as an option was added in 3.8, but there's really no need to check. If the query returns 17, the user is definitely running 3.8+. If the user is running pre-3.8 then the query can never yield the value 17 (because it cannot be set), but the script engine won't argue about a string check.

**Comment by Ruyk** [comment]
Looks good to me then!

**Pull request merged**